src/multi.c

```diff
@@ -264,16 +264,72 @@ void execCommand(client *c) {
  * Also every client contains a list of WATCHed keys so that's possible to
  * un-watch such keys when the client is freed or when UNWATCH is called. */
 
-/* In the client->watched_keys list we need to use watchedKey structures
- * as in order to identify a key in Redis we need both the key name and the
- * DB. This struct is also referenced from db->watched_keys dict, where the
- * values are lists of watchedKey pointers. */
+/* In the client->watched_keys list (keys watched by the client) we need to use
+ * watchedKey structures as in order to identify a key in Redis we need both the
+ * key name and the DB. There is also a listNode embed in this structure. Nodes
+ * embedded in watchedKey structures with a same key make up the list of clients
+ * watching that key. We point embedded node->value back to the list. */
```

Comment:
I think this description is better (your recent version didn't mention db->watched_keys.
i think that given this, we can drop the fancy diagram which feels to me a bit too verbose (we don't have such diagrams for other parts of redis with similar complexity)
WDYT?
```suggestion
/* The watchedKey struct is included in two lists: the client->watched_keys list,
 * and db->watched_keys dict (each value in that dict is a list of watchedKey structs).
 * The list in the client struct is a plain list, where each node's value is a pointer to a watchedKey.
 * The list in the db db->watched_keys is different, the listnode member that's embedded in this struct
 * is the node in the dict. And the value inside that listnode is a pointer to the that list, and we can use
 * struct member offset math to get from the listnode to the watchedKey struct.
 * This is done to avoid the need for listSearchKey and dictFind when we remove from the list. */

Reply:
Updated, these comments looked better!